// react-contentful/Query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withContentful = _interopRequireDefault(require("./withContentful"));

var _helpers = require("./helpers");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Query = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(Query, _Component);

  var _super = _createSuper(Query);

  function Query(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    var data = (0, _helpers.checkCache)(props);
    _this.state = {
      fetched: data ? true : false,
      loading: false,
      error: null,
      data: data
    };

    if (data) {
      props.onLoad(_this.state);
    }

    return _this;
  }

  var _proto = Query.prototype;

  _proto.componentDidMount = function componentDidMount() {
    if (!this.state.data) {
      this.requestContentfulData();
    }
  };

  _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
    var _this2 = this;

    if (JSON.stringify(this.props) !== JSON.stringify(prevProps)) {
      this.setState({
        fetched: false
      }, function () {
        _this2.requestContentfulData();
      });
    }
  };

  _proto.requestContentfulData = function requestContentfulData() {
    var _this3 = this;

    var _this$props = this.props,
        parser = _this$props.parser,
        onRequest = _this$props.onRequest,
        onLoad = _this$props.onLoad,
        onError = _this$props.onError;
    (0, _helpers.validateRequestRequirements)(this.props).then(function () {
      _this3.setState({
        error: null,
        loading: true
      }, /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                onRequest(_this3.state);
                (0, _helpers.fetchData)(_this3.props).then(function (response) {
                  _this3.setState({
                    data: parser(response, _this3.props),
                    fetched: true,
                    loading: false
                  }, function () {
                    onLoad(_this3.state);
                  });
                });

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      })));
    })["catch"](function (error) {
      _this3.setState({
        error: error,
        fetched: true,
        loading: false
      }, function () {
        onError(_this3.state);
      });
    });
  };

  _proto.getQueryResult = function getQueryResult() {
    return this.state;
  };

  _proto.render = function render() {
    var _this4 = this;

    var _this$props2 = this.props,
        children = _this$props2.children,
        contentful = _this$props2.contentful;

    var finish = function finish() {
      return children(_this4.getQueryResult());
    };

    if (contentful && contentful.renderPromises) {
      return contentful.renderPromises.addQueryPromise(this, finish);
    }

    return finish();
  };

  return Query;
}(_react.Component);

Query.propTypes = {
  children: _propTypes["default"].func,
  contentType: _propTypes["default"].string,
  id: _propTypes["default"].string,
  include: _propTypes["default"].number,
  query: _propTypes["default"].object,
  parser: _propTypes["default"].func,
  skip: _propTypes["default"].bool,
  onError: _propTypes["default"].func,
  onLoad: _propTypes["default"].func,
  onRequest: _propTypes["default"].func
};
Query.defaultProps = {
  children: function children(_ref2) {
    var data = _ref2.data,
        error = _ref2.error,
        fetched = _ref2.fetched,
        loading = _ref2.loading;
    return null;
  },
  include: 10,
  query: {},
  skip: false,
  parser: function parser(data, props) {
    return data;
  },
  onError: function onError() {},
  onLoad: function onLoad() {},
  onRequest: function onRequest() {}
};

var _default = (0, _withContentful["default"])(Query);

exports["default"] = _default;
