// react-contentful/es/ContentfulProvider.js
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import warning from 'warning';
import ContentfulContext from './ContentfulContext';

var ContentfulProvider = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ContentfulProvider, _Component);

  var _super = _createSuper(ContentfulProvider);

  function ContentfulProvider(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    var client = props.client,
        locale = props.locale,
        renderPromises = props.renderPromises;
    _this.state = {
      client: client,
      locale: locale,
      renderPromises: renderPromises
    };
    return _this;
  }

  var _proto = ContentfulProvider.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        context = _this$props.context;
    var Context = context || ContentfulContext;
    warning(this.state.client, 'No `client` specified on <ContentfulProvider />');
    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: this.state
    }, children);
  };

  return ContentfulProvider;
}(Component);

ContentfulProvider.propTypes = {
  children: PropTypes.any,
  client: PropTypes.object.isRequired,
  context: PropTypes.object,
  locale: PropTypes.string,
  renderPromises: PropTypes.object
};
ContentfulProvider.defaultProps = {
  children: null,
  context: null,
  locale: 'en-US',
  renderPromises: null
};
export default ContentfulProvider;
